### PR TITLE
Create compounds_review

### PR DESCRIPTION
**Pull request by EricaBiagetti**
In the new file **compounds_review_1-40** (the numbers indicate the sentences reviewed so far) I propose a new way of treating compounds within the Sanskrit treebank.
Although UD guidelines require compounds not to be split, confrontations with Sanskrit scholars have lead me to analyze every compund as a multi-word token. In the same way are treated words which are united due to sandhi, having thus no syntactic or semantic connection between eachother. 
In order to distinguish between sandhi forms and compounds, the following guidelines will be followed:

- **sandhi forms** are treated as multi-word tokens, with the first line containing `ID`,`FORM` and `MISC`, this one filled up with `Translit` and `Gloss` of the form itself. Every single element is then analyzed as usual, with its morphological features (`FEATS`) and the `MISC` filled up with `Translit`,`LTranslit` (Lemma Transliteration) and `Gloss`.
```
16-18	ग्राह्यमपास्य	_	_	_	_	_	_	_	Translit=grāhyamapāsya|Gloss=to-be-taken-having-rejected-outside
16	ग्राह्यम्	ग्रह्	ADJ	_	Case=Nom|Gender=Neut|Number=Sing|Tense=Fut|VerbForm=Part|Voice=Pass	1	parataxis	_	Translit=grāhyam|LTranslit=grah|Gloss=to-be-taken
17	अप	अप	ADV	_	_	18	advmod	_	Translit=apa|LTranslit=apa|Gloss=outside
18	अस्य	अस्	VERB	_	VerbForm=Conv	16	advcl	_	Translit=asya|LTranslit=as|Gloss=having-rejected
```

- **compounds**  first line is similar to the one above but, in order to give the idea of an (often) non-compositional meaning, `Gloss` is given its value between brackets.
Since the first(s) element of a compound lack nominal inflection,`FEATS` has a language-specific feature `Compund=Yes`. This is to specify that this element do not stand alone and make it possible to give verbal features (such as `Mood)` as well as pronominal ones (`PronType`) and others (`Polarity`) which are not lost in compounds first element(s).

```
1-3	प्रभूतकालज्ञेयानि	_	_	_	_	_	_	_	Translit=prabhūtakālajñeyāni|Gloss=(who-will-be-known-with-time-spent)
1	प्रभूत	प्र-भूत	ADJ	_	Compound=Yes|Tense=Past|VerbForm=Part|Voice=Pass	2	amod	_	Translit=prabhūta|LTranslit=pra-bhūta|Gloss=spent
2	काल	काल	NOUN	_	Compound=Yes	3	obl	_	LId=काल-१|Translit=kāla|LTranslit=kāla|Gloss=time
3	ज्ञेयानि	ज्ञेय	ADJ	_	Case=Nom|Gender=Neut|Number=Plur|Tense=Fut|VerbForm=Part|Voice=Pass	0	root	_ Translit=jñeyāni|LTranslit=jñeya|Gloss=who-will-be-known
```

- in case words that do not take part in a compound surface as a single token with it due to sandhi, `Gloss` will be given the compund's meaning within brackets and the meaning of the remaining words outside of it.
```
3-6	मदनुग्रहार्थमेतान्	_	_	_	_	_	_	_	Translit=madanugrahārthametān|Gloss=(for-reason-of-a-favor-to-me)-these
3	मत्	अस्मद्	PRON	_	Compound=Yes|Person=1|PronType=Prs	4	nmod:poss	_	Translit=mat|LTranslit=asmad|Gloss=me
4	अनुग्रह	अनुग्रह	NOUN	_	Compound=Yes	5	nmod	_	Translit=anugraha|LTranslit=anugraha|Gloss=favor
5	अर्थम्	अर्थ	NOUN	_	Case=Acc|Gender=Masc|Number=Sing	13	obl	_	Translit=artham|LTranslit=artha|Gloss=for-the-reason
6	एतान्	एतद्	PRON	_	Case=Acc|Gender=Masc|Number=Plur|PronType=Dem	13	obj	_	Translit=etān|LTranslit=etad|Gloss=these
```

With this _pull  request_ I only gave the main guidelines that has been followed during the review of the first 40 sentences of the treebank. Other cases should be discussed, such as those in which I had to modify the text because it contained compound surfacing as two tokens. The `DEPREL` `compound` has also been removed, so that grammatical relations between elements of a compound can be analized.
This approach is meant to make the Sanskrit treebank useful for Sanskrit scholars beside computational linguists. I wuold be pleased to receive any kind of feedbacks about it.

### CODE REVIEW

**Review by dan-zeman — COMMENTED** [review]
The proposal sounds good to me. Making the treebank more useful to Sanskrit scholars is definitely one of the objectives that we should not ignore.

Technical note: In my copy, I overwrote sentences 1-40 in sa_ufal-ud-test.conllu by the contents of your compounds_review_1-40, in order to see the diff. But I see that many differences are actually in the encoding of the dotted characters in the transliteration. I will adapt my software to render them as single characters where possible.

**Comment by dan-zeman** [comment]
I noticed that you modified tokenization in the # translit line without doing the same change in the # text line. That is bad (although it does not make the file invalid from the UD perspective) because then the transliteration no longer corresponds to the original text. I noticed one case but I did not check whether and how many others there are. Here, _śāsanaśatena_ were originally two words, _śāsana śatena_:

<pre>
# sent_id = panc0.s34
# text = नाहं विद्याविक्रयं करोमि शासन शतेन।
# translit = nāhaṁ vidyāvikrayaṁ karomi śāsanaśatena.
</pre>

If translit is changed, then the corresponding space in text should be also erased: शासनशतेन।

**Comment by dan-zeman** [comment]
Please check that the data is valid.

Use the official validating script, validate.py, from the tools repository. At present there are errors in compounds_review:

<pre>
[Line                   20]: Violation of guidelines: relation conj must go left-to-right
[Line                   21]: Violation of guidelines: relation flat must go left-to-right
[Line                   32]: Extra characters at the end of the text attribute, not accounted for in the FORM fields: '।'
[Line                   42]: Violation of guidelines: relation conj must go left-to-right
[Line                   43]: Violation of guidelines: relation conj must go left-to-right
[Line                   50]: Violation of guidelines: relation conj must go left-to-right
[Line                   53]: Violation of guidelines: relation conj must go left-to-right
[Line                   54]: The line has 9 columns, but 10 are expected.
[Line                   54]: White space not allowed in the XPOSTAG column: '_       _'
[Line                   54]: Spurious morphological feature: '15'. Should be of the form attribute=value and must start with [A-Z0-9] and only contain [A-Za-z0-9].
[Line                   54]: Invalid DEPREL value _
[Line                   54]: Failed for parse DEPS: Translit=adadhi|LTranslit=adadhi|Gloss=oceans
[Line                   54]: Unknown UD DEPREL: _
[Line                   54]: Malformed head:deprel pair 'Translit=adadhi'
[Line                   54]: Malformed head:deprel pair 'LTranslit=adadhi'
[Line                   54]: Malformed head:deprel pair 'Gloss=oceans'
[Line                   55]: Violation of guidelines: relation conj must go left-to-right
[Line                   59]: Violation of guidelines: relation conj must go left-to-right
[Line                   62]: Undefined ID in HEAD: conj
[Line                   62]: Failed for parse DEPS: Translit=adadhi|LTranslit=adadhi|Gloss=oceans
[Line                   62]: Failed to parse DEPS: Translit=adadhi|LTranslit=adadhi|Gloss=oceans
[Tree number 4 on line 38]: Non-integer head for word ID 13
[Tree number 4 on line 38]: Non-tree structure. Words 13 are not reachable from the root 0.
[Line                   62]: Exception caught!
*** FAILED *** with 24 errors
Format errors: 7
Metadata errors: 1
Morpho errors: 1
Syntax errors: 15
</pre>

**Comment by EricaBiagetti** [comment]
I've noticed that some syntax errors are due to the annotation of coordinative compounds. For example, the error `[Line n]: Violation of guidelines: relation conj must go left-to-right` appears in a sentence such as: 
```
# newpar id = panc0.p2
# sent_id = panc0.s2
# text = ओं नमः श्रीशारदागणपतिगुरुभ्यः।
# translit = oṁ namaḥ śrīśāradāgaṇapatigurubhyaḥ.
1	ओं	ओं	INTJ	_	_	2	discourse	_	Translit=oṁ|LTranslit=oṁ|Gloss=oh
2	नमः	नमस्	NOUN	_	Case=Nom|Gender=Neut|Number=Sing	0	root	_	Translit=namaḥ|LTranslit=namas|Gloss=glory
3-6	श्रीशारदागणपतिगुरुभ्यः	_	_	_	_	_	_	_	SpaceAfter=No|Translit=śrīśāradāgaṇapatigurubhyah|Gloss=(to-the-lords-honorable-Śāradā-and-Gaṇapati)
3	श्री	श्री	ADJ	_	Compound=Yes	4	amod	_	Translit=śrī|LTranslit=śrī|Gloss=honorable
4	शारदा	शारदा	PROPN	_	Compound=Yes	6	conj	_	Translit=śāradā|LTranslit=śāradā|Gloss=Śāradā
5	गणपति	गणपति	PROPN	_	Compound=Yes	6	flat	_	Translit=gaṇapati|LTranslit=gaṇapati|Gloss=Gaṇapati
6	गुरुभ्यः	गुरु	NOUN	_	Case=Dat|Gender=Masc|Number=Plur	2	nmod	_	Translit=gurubhyaḥ|LTranslit=guru|Gloss=lords
7	।	।	PUNCT	_	_	2	punct	_	Translit=.|LTranslit=.|Gloss=.
```

Such compounds lack a head and I've decided to treat the last element as the head with the previous ones depending on it as `conj`, in this way all compounds would have been regarded as having the last element as the head.
Since the validation doesn't permit a `conj `deprel which goes from right to left, I then have to review them and annotate all non-first elements as conjuncts of the first. From a theoretical point of view though, this could look not appropriate because it's the last element which carries morphological inflection and determines the grammatical relation with respect to other words in the sentences.

Should I procede this way?

**Comment by martinpopel** [comment]
I don't see any `conj` in your example.
In UD, coordination is considered a paratactic structure with no linguistic head, but for consistency the first conjunct is always the head no matter whether it carries any morphological inflection etc. Of course, it needs to be decided for each language which all constructions are considered a coordination, e.g. "A and B" vs. "A with B" vs. "A accompanying B".

**Comment by dan-zeman** [comment]
@martinpopel is right. The relation `conj` is not a dependency in the strict sense. Therefore it goes always left-to-right in UD, regardless who carries morphological inflection.

**Comment by dan-zeman** [comment]
I have moved the compounds_review file to the original name, sa_ufal-ud-test.conllu. Please modify directly this file from now on. I am not merging the pull request to the dev branch yet, as the file is not valid at the moment.

**Comment by EricaBiagetti** [comment]
I've corrected the example. The issue also pertains the relation `flat`, but I can correct all of them and annotate the first element as head and the following as its `conj`.

File: sa_ufal-ud-test.conllu [hunk 1]
```diff
@@ -12,14 +12,14 @@
 # newpar id = panc0.p2
 # sent_id = panc0.s2
 # text = ओं नमः श्रीशारदागणपतिगुरुभ्यः।
-# translit = oṁ namaḥ śrīśāradāgaṇapatigurubhyaḥ.
-1	ओं	ओं	INTJ	_	_	2	discourse	_	Translit=oṁ|LTranslit=oṁ|Gloss=oh
-2	नमः	नमस्	NOUN	_	Case=Nom|Gender=Neut|Number=Sing	0	root	_	Translit=namaḥ|LTranslit=namas|Gloss=glory
-3-6	श्रीशारदागणपतिगुरुभ्यः	_	_	_	_	_	_	_	SpaceAfter=No|Translit=śrīśāradāgaṇapatigurubhyah|Gloss=(to-the-lords-honorable-Śāradā-and-Gaṇapati)
+# translit = oṁ namaḥ śrīśāradāgaṇapatigurubhyaḥ.
```
Comment (dan-zeman) [inline sa_ufal-ud-test.conllu]: Why have you changed the encoding of the accented characters in the transliteration?

File: sa_ufal-ud-test.conllu [hunk 2]
```diff
@@ -563,7 +563,7 @@
 7	वात्स्यायनादीनि	वात्स्यायनादि	PROPN	_	Case=Acc|Gender=Neut|Number=Plur	6	appos	_	SpaceAfter=No|Translit=vātsyāyanādīni|LTranslit=vātsyāyanādi|Gloss=works-by-Vātsyāyana
 8	।	।	PUNCT	_	_	2	punct	_	Translit=.|LTranslit=.|Gloss=.
 
-# sent_id = panc0.s18
+# # sent_id = panc0.s18
 # text = एवञ्च ततो धर्मार्थकामशास्त्राणि ज्ञायन्ते।
```
Comment (dan-zeman) [inline sa_ufal-ud-test.conllu]: This extra `#` also means that the mandatory `sent_id` comment is not recognized.

**Review by dan-zeman — CHANGES_REQUESTED** [review]
The resulting file is not valid:

<pre>
[Line                   32]: Extra characters at the end of the text attribute, not accounted for in the FORM fields: '।'
[Line                   54]: Trailing whitespace not allowed in column XPOSTAG
[Line                   54]: Initial whitespace not allowed in column FEATS
[Line                   54]: White space not allowed in the XPOSTAG column: '_   '
[Line                   54]: White space not allowed in the FEATS column: '    _'
[Line                   54]: Spurious morphological feature: '    _'. Should be of the form attribute=value and must start with [A-Z0-9] and only contain [A-Za-z0-9].
[Tree number 4 on line 38]: Mismatch between the text attribute and the FORM field. Form is '।' but text is ' ।...'
[Line                   62]: Extra characters at the end of the text attribute, not accounted for in the FORM fields: ' ।'
[Line                   74]: Unknown attribute-value pair Number=Plural
[Line                   93]: Missing the sent_id attribute.
[Tree number 5 on line 67]: Mismatch between the text attribute and the FORM field. Form is '॥' but text is ' ॥...'
[Line                   93]: Extra characters at the end of the text attribute, not accounted for in the FORM fields: ' ॥'
[Line                   118]: Unknown attribute-value pair Compund=Yes
[Tree number 7 on line 111]: Mismatch between the text attribute and the FORM field. Form is '॥' but text is ' ॥...'
[Line                   121]: Extra characters at the end of the text attribute, not accounted for in the FORM fields: ' ॥'
[Tree number 13 on line 240]: Mismatch between the text attribute and the FORM field. Form is 'शास्तविमुखान्' but text is 'शास्त्रविमुखान् समालोक्य स राजा स...'
[Tree number 13 on line 240]: Mismatch between the text attribute and the FORM field. Form is 'समालोक्य' but text is 'शास्त्रविमुखान् समालोक्य स र...'
[Tree number 13 on line 240]: Mismatch between the text attribute and the FORM field. Form is 'स' but text is 'शास्त्रविमुखान् समालो...'
[Tree number 13 on line 240]: Mismatch between the text attribute and the FORM field. Form is 'राजा' but text is 'शास्त्रविमुखान् समालोक्य...'
[Tree number 13 on line 240]: Mismatch between the text attribute and the FORM field. Form is 'सचिवान्' but text is 'शास्त्रविमुखान् समालोक्य स ...'
[Tree number 13 on line 240]: Mismatch between the text attribute and the FORM field. Form is 'आहूय' but text is 'शास्त्रविमुखान् समालोक्य...'
[Tree number 13 on line 240]: Mismatch between the text attribute and the FORM field. Form is 'प्रोवाच' but text is 'शास्त्रविमुखान् समालोक्य स ...'
[Tree number 13 on line 240]: Mismatch between the text attribute and the FORM field. Form is '।' but text is 'शास्त्रविमुखान् समालो...'
[Line                   252]: Extra characters at the end of the text attribute, not accounted for in the FORM fields: 'शास्त्रविमुखान् समालोक्य स राजा सचिवान् आहूय प्रोवाच ।'
[Tree number 22 on line 466]: Mismatch between the text attribute and the FORM field. Form is '।' but text is ' ।...'
[Line                   481]: Extra characters at the end of the text attribute, not accounted for in the FORM fields: ' ।'
...suppressing further errors regarding Metadata
[Line                   626]: The line has 9 columns, but 10 are expected.
[Line                   626]: A token line must have '_' in the column DEPS. Now: 'SpaceAfter=No|Translit=jīvitavyaviṣayaḥ|Gloss=(duration-of-life)'.
[Line                   630]: Exception caught!
*** FAILED *** with 30 errors
Format errors: 7
Metadata errors: 20
Morpho errors: 3
Traceback (most recent call last):
  File "/net/work/people/zeman/unidep/tools/validate.py", line 798, in &lt;module&gt;
    validate(inp,out,args,tagsets,known_sent_ids)
  File "/net/work/people/zeman/unidep/tools/validate.py", line 673, in validate
    validate_text_meta(comments,tree)
  File "/net/work/people/zeman/unidep/tools/validate.py", line 179, in validate_text_meta
    if u"NoSpaceAfter=Yes" in cols[MISC]: #I leave this without the split("|") to catch all
IndexError: list index out of range
</pre>

**Comment by EricaBiagetti** [comment]
Dear Mr. Zeman, right now I'm having problems with my text editor, but I'll solve them as soon as possible.

**Pull request merged**